sycl/doc/design/KernelFusionJIT.md

```diff
@@ -158,7 +158,115 @@ The metadata is attached to a function that will become the fused kernel:
 
+### Fusing kernels with different nd-ranges
+
+This section explains actions performed by the kernel fusion JIT compiler when fusing kernels with different nd-ranges. Throughout this section, we refer to "work-item components". A comprehensive list of these components mentioned in this document is:
+
+- `global_size`
+- `local_size`
+- `num_work_groups`
+- `global_id`
+- `local_id`
+- `group_id`
+- `global_offset`
+
+Meaning of each of these is self-explainatory for the SYCL user.
+
+#### Restrictions
+
+Following kernel fusion principles, SYCL constraints and technical decisions, some basic constraints are set for valid combinations of nd-ranges:
+
+1. The fused kernel should perform no more visible work than the union of the unfused kernels;
+2. The fused kernel should perform no less visible work than the union of the unfused kernels;
+3. If two work items belong to the same work-group in one of the unfused grids, they must also belong to the same work-group in the fused grid;
+4. Either none or all of the work-items of a work-group must execute barriers inserted by the kernel fusion process;
+5. The fused kernel must not launch more work-items than the maximum number of work-items launched by the original kernels.
+6. All work-groups will be the same size, [as per the SYCL 2020 rev 7. 3.9.4](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_work_group_data_parallel_kernels).
+7. `global_id(i) = group_id(i) * local_size(i) + local_id(i)` [as per OpenCL 3.0 3.2.1](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_mapping_work_items_onto_an_ndrange).
+8. A work-item will have the same global linear ID in the fused grid as in the unfused grid;
+9. All the fused nd-ranges must have the same offset.
+
+These restrictions can be simplified to:
+
+- No two local sizes specified by the nd-ranges will be different;
+- No global id remapping is needed ([see](#work-item-remapping)) or all input offsets are 0;
+- All the fused nd-ranges must have the same offset.
+
+As we can see, there is no restrictions in the number of dimensions or global sizes of the input nd-ranges.
+
+#### Work-item remapping
+
+Work-item remapping is performed at the input kernel level, i.e., a different remapping is performed for each input kernel, as different input nd-ranges will result in different remappings.
+
+This remapping consists on an inter-procedural pass replacing each built-in querying components of a work-item, e.g., the global id or the local size, with a JIT-generated value.
+
+First of all, work-item remapping will always be performed when the list of input nd-ranges is heterogeneous. Additional remapping conditions are present for the following work-item components. For each input kernel:
+
+- `num_work_groups` and `local_size`: Only performed if the input nd-range has an explicit local size, may result in better performance, as this replaces built-in calls with constants;
```

Comment:
`global_offset` also falls into this category.

Reply:
`global_size` and `global_offset` are always remapped when nd-ranges are heterogeneous because we always have the constant values there ([SPIR-V](https://github.com/intel/llvm/blob/1a8b273d026f1617f7afd63d668513a763894b9a/sycl-fusion/passes/target/TargetFusionInfo.cpp#L285) and [CUDA](https://github.com/intel/llvm/blob/1a8b273d026f1617f7afd63d668513a763894b9a/sycl-fusion/passes/target/TargetFusionInfo.cpp#L512)).